Python/ceval.c

```diff
@@ -4223,6 +4225,69 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, InterpreterFrame *frame, int thr
             DISPATCH();
         }
 
+        case TARGET(LOAD_METHOD_ADAPTIVE): {
+            assert(cframe.use_tracing == 0);
+            SpecializedCacheEntry *cache = GET_CACHE();
+            if (cache->adaptive.counter == 0) {
+                PyObject *owner = TOP();
+                PyObject *name = GETITEM(names, cache->adaptive.original_oparg);
+                next_instr--;
+                if (_Py_Specialize_LoadMethod(owner, next_instr, name, cache) < 0) {
+                    goto error;
+                }
+                DISPATCH();
+            }
+            else {
+                STAT_INC(LOAD_METHOD, deferred);
+                cache->adaptive.counter--;
+                oparg = cache->adaptive.original_oparg;
+                STAT_DEC(LOAD_METHOD, unquickened);
+                JUMP_TO_INSTRUCTION(LOAD_METHOD);
+            }            
+        }
+
+        case TARGET(LOAD_METHOD_CACHED): {
+            /* Designed to work in tandem with CALL_METHOD. */
+            assert(cframe.use_tracing == 0);
+            PyObject *self = TOP();
+            PyTypeObject *self_cls = Py_TYPE(self);
+            PyObject *res = NULL;
+            SpecializedCacheEntry *caches = GET_CACHE();
+            _PyAdaptiveEntry *cache0 = &caches[0].adaptive;
+            _PyLoadMethodCache *cache1 = &caches[-1].load_method;
+            _PyLoadMethodCache *cache2 = &caches[-2].load_method;
+
+            DEOPT_IF(self_cls->tp_version_tag != cache1->attr.tp_version, LOAD_METHOD);
+            assert(cache1->attr.dk_version_or_hint != 0);
+            assert(cache1->attr.tp_version != 0);
+            assert(self_cls->tp_dictoffset >= 0);
+            assert(Py_TYPE(self_cls)->tp_dictoffset > 0);
+            
+            PyObject *name = GETITEM(names, cache0->original_oparg);
+            // inline version of _PyObject_GetDictPtr for offset >= 0
+            PyObject **dictptr = self_cls->tp_dictoffset != 0 ?
+                (PyObject **) ((char *)self + self_cls->tp_dictoffset) : NULL;
+            PyDictObject *dict = NULL;
+
+            // ensure self.__dict__ didn't modify keys
+            if (dictptr != NULL && *dictptr != NULL) {
```

Comment:
Just deopt if the `dict` is NULL. The `tp_version` check guarantees that `tp_dictoffset` is as expected.
```C
PyObject *dict = *(PyObject **) ((char *)self + self_cls->tp_dictoffset);
DEOPT_IF(dict == NULL, LOAD_METHOD);
DEOPT_IF(dict->ma_keys->dk_version != cache1->attr.dk_version_or_hint, LOAD_METHOD);
```

Reply:
I'm being stupid. You don't need to deopt if `dict==NULL`.
```C
PyObject *dict = *(PyObject **) ((char *)self + self_cls->tp_dictoffset);
DEOPT_IF(dict != NULL && dict->ma_keys->dk_version != cache1->dk_version_or_hint, LOAD_METHOD);
```

Reply:
Thank you! That's pretty smart, we managed to avoid one more branch :).

``PyObject *dict = *(PyObject **) ((char *)self + self_cls->tp_dictoffset);`` segfaults, because ``self_cls->tp_dictoffset`` can be ``0``. So I left the old way with the ternary operator in.

Reply:
`tp_dictoffset` cannot change. Do you not check for `tp_dictoffset == 0` when specializing?

Reply:
``tp_dictoffset == 0`` is fine, because that means the object has no dict.